File: frei0r/mediafx_filter/examples/shift.rs
// Copyright (C) 2025 Andrew Wason
// SPDX-License-Identifier: GPL-3.0-or-later

use mediafx::frameserver::client;

// XXX need to forward params to client
const XSHIFT: f32 = 0.5;
const YSHIFT: f32 = 0.25;

fn filter_frame(frame_client: client::FrameClient) -> client::FrameClient {
    let size = frame_client.render_size();
    let mut request = frame_client.request_render().unwrap();

    let xshift = (XSHIFT * size.width() as f32) as u32;
    let yshift = (YSHIFT * size.height() as f32) as u32;
    for dy in 0..size.height() {
        for dx in 0..size.width() {
            let sy = (dy + yshift) % size.height();
            let sx = (dx + xshift) % size.width();
            let dest_index = (dy * size.width() + dx) as usize * 4;
            let source_index = (sy * size.width() + sx) as usize * 4;
            for channel in 0..4 {
                request.get_rendered_frame_mut()[dest_index + channel] =
                    request.get_source_frame(0).unwrap()[source_index + channel];
            }
        }
    }

    request.render_complete().unwrap()
}

fn main() {
    let mut frame_client = client::FrameClient::new().unwrap();
    loop {
        frame_client = filter_frame(frame_client);
    }
}

File: clients/mediafx-py/python/tests/shift.py
#!/usr/bin/env python
# Copyright (C) 2025 Andrew Wason
# SPDX-License-Identifier: GPL-3.0-or-later

import struct
import sys
from mediafx import MediaFX

XSHIFT = 0.5
YSHIFT = 0.25

def main():
    client = MediaFX()
    width, height = client.frame_size
    pixels = width * height
    rendered_frame = bytearray(client.frame_bytecount)
    frames = [bytearray(client.frame_bytecount)]

    while True:
        time = client.render_begin(frames)

        xshift = int(XSHIFT * width);
        yshift = int(YSHIFT * height);
        for dy in range(0, height):
            for dx in range(0, width):
                sy = (dy + yshift) % height
                sx = (dx + xshift) % width
                dest_index = (dy * width + dx) * 4
                source_index = (sy * width + sx) * 4
                for channel in range(0, 4):
                    rendered_frame[dest_index + channel] = frames[0][source_index + channel]

        client.render_finish(rendered_frame)

if __name__ == "__main__":
    main()
